src/swe/swe_operators.c

```diff
@@ -0,0 +1,364 @@
+
+      // create an element restriction for geometric factors that convert
+      // fluxes to cell states
+      CeedInt g_strides[] = {num_comp_geom, 1, num_comp_geom};
+      CeedElemRestrictionCreateStrided(ceed, num_edges, 1, num_comp_geom, num_edges * num_comp_geom, g_strides, &restrict_geom);
+      CeedElemRestrictionCreateVector(restrict_geom, &geom, NULL);
+      CeedVectorSetValue(geom, 0.0);
+
+      // create an element restriction for accumulated fluxes
+      CeedInt f_strides[] = {num_comp, 1, num_comp};
+      CeedElemRestrictionCreateStrided(ceed, num_edges, 1, num_comp, num_edges * num_comp, f_strides, &restrict_flux);
+      CeedElemRestrictionCreateVector(restrict_flux, &flux, NULL);
+      CeedVectorSetValue(flux, 0.0);
+
+      // create element restrictions for left and right input/output states,
+      // populate offsets for these states, and set the (invariant)
+      // geometric parameters
+      CeedInt *offset_l, *offset_r;
+      PetscCall(PetscMalloc2(num_edges, &offset_l, num_edges, &offset_r));
+      CeedScalar(*g)[4];
+      CeedVectorGetArray(geom, CEED_MEM_HOST, (CeedScalar **)&g);
+      for (CeedInt e = 0, oe = 0; e < mesh->num_internal_edges; e++) {
+        PetscInt iedge = edges->internal_edge_ids[e];
+        if (!edges->is_owned[iedge]) continue;
+        PetscInt l   = edges->cell_ids[2 * iedge];
+        PetscInt r   = edges->cell_ids[2 * iedge + 1];
+        offset_l[oe] = l * num_comp;
+        offset_r[oe] = r * num_comp;
+
+        g[oe][0] = edges->sn[iedge];
+        g[oe][1] = edges->cn[iedge];
+        g[oe][2] = -edges->lengths[iedge] / cells->areas[l];
+        g[oe][3] = edges->lengths[iedge] / cells->areas[r];
+        oe++;
+      }
+      CeedVectorRestoreArray(geom, (CeedScalar **)&g);
+
+      // create element restrictions for left and right cell states
+      CeedElemRestrictionCreate(ceed, num_edges, 1, num_comp, 1, mesh->num_cells * num_comp, CEED_MEM_HOST, CEED_COPY_VALUES, offset_l, &restrict_l);
+      CeedElemRestrictionCreate(ceed, num_edges, 1, num_comp, 1, mesh->num_cells * num_comp, CEED_MEM_HOST, CEED_COPY_VALUES, offset_r, &restrict_r);
+      PetscCall(PetscFree2(offset_l, offset_r));
+      if (0) {
+        CeedElemRestrictionView(restrict_l, stdout);
+        CeedElemRestrictionView(restrict_r, stdout);
+      }
+    }
+
+    {
+      CeedOperator op;
+      CeedOperatorCreate(ceed, qf, NULL, NULL, &op);
+      CeedOperatorSetField(op, "geom", restrict_geom, CEED_BASIS_COLLOCATED, geom);
+      CeedOperatorSetField(op, "q_left", restrict_l, CEED_BASIS_COLLOCATED, CEED_VECTOR_ACTIVE);
+      CeedOperatorSetField(op, "q_right", restrict_r, CEED_BASIS_COLLOCATED, CEED_VECTOR_ACTIVE);
+      CeedOperatorSetField(op, "cell_left", restrict_l, CEED_BASIS_COLLOCATED, CEED_VECTOR_ACTIVE);
+      CeedOperatorSetField(op, "cell_right", restrict_r, CEED_BASIS_COLLOCATED, CEED_VECTOR_ACTIVE);
+      CeedOperatorSetField(op, "flux", restrict_flux, CEED_BASIS_COLLOCATED, flux);
```

Comment:
Since we specified a vector (`flux` as the fifth argument) instead of `CEED_VECTOR_ACTIVE`, it becomes a "passive field". The `geom` is also a passive field except `geom` is an input field (via [`CeedQFunctionAddInput#L47`](https://github.com/RDycore/RDycore/pull/90/files#diff-cebc7026ff0766eb019ae159a8319ca18c5fa7d87b58278c9ed9ec7957ac2e53R47)), while `flux` is an output field (via [`CeedQFunctionAddOutput#L52`](https://github.com/RDycore/RDycore/pull/90/files#diff-cebc7026ff0766eb019ae159a8319ca18c5fa7d87b58278c9ed9ec7957ac2e53R52)). Did I get this correct?

Reply:
Yes, that is how I understand it.